File: practice-javascript/checkout.js
class Product {
    constructor(id, name, price) {
      this.id = id;
      this.name = name;
      this.price = price;
    }
  }
  
  class ShoppingCart {
    constructor() {
      this.items = [];
    }
  
    async addItem(product, quantity) {
      // Simulate an asynchronous operation (e.g. fetching product data from a server)
      const productData = await fetchProductData(product.id);
  
      if (!productData) {
        throw new Error(`Product with ID ${product.id} not found.`);
      }
  
      const item = {
        product: product,
        quantity: quantity,
      };
  
      this.items.push(item);
      console.log(`${quantity} ${product.name}(s) added to the cart.`);
    }
  
    removeItem(productId) {
      const itemIndex = this.items.findIndex(item => item.product.id === productId);
      if (itemIndex !== -1) {
        this.items.splice(itemIndex, 1);
        console.log(`Product with ID ${productId} removed from the cart.`);
      } else {
        console.log(`Product with ID ${productId} not found in the cart.`);
      }
    }
  
    calculateTotal() {
      return this.items.reduce((total, item) => total + item.product.price * item.quantity, 0);
    }
  
    async checkout() {
      const total = this.calculateTotal();
      console.log(`Checking out... Total: $${total.toFixed(2)}`);
  
      // Simulate an asynchronous payment process
      await processPayment(total);
  
      console.log('Payment successful. Thank you for your purchase!');
      this.items = []; // Empty the cart after successful checkout
    }
  }
  
  async function fetchProductData(productId) {
    // Simulate fetching product data from a server
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        const products = [
          new Product(1, 'Laptop', 999.99),
          new Product(2, 'Phone', 599.99),
          new Product(3, 'Headphones', 99.99),
        ];
  
        const product = products.find(p => p.id === productId);
        resolve(product);
      }, 1000);
    });
  }
  
  async function processPayment(total) {
    // Simulate a payment process
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        if (Math.random() < 0.9) {
          resolve();
        } else {
          reject('Payment failed. Please try again.');
        }
      }, 1500);
    });
  }
  
  const cart = new ShoppingCart();
  
  (async () => {
    try {
      await cart.addItem(new Product(1, 'Laptop', 1), 2);
      await cart.addItem(new Product(2, 'Phone', 2), 1);
      cart.removeItem(3);
      await cart.checkout();
    } catch (error) {
      console.error(error);
    }
  })();  

File: practice-python/checkout.py
class Product:
    def __init__(self, product_id, name, price):
        self.product_id = product_id
        self.name = name
        self.price = price

class ShoppingCart:
    def __init__(self):
        self.items = []

    def add_item(self, product, quantity):
        """Add a product and its quantity to the shopping cart."""
        self.items.append({"product": product, "quantity": quantity})
        print(f"Added {quantity} {product.name}(s) to the cart.")

    def remove_item(self, product_id, quantity):
        """Remove a certain quantity of a product from the shopping cart."""
        for item in self.items:
            if item["product"].product_id == product_id:
                if item["quantity"] >= quantity:
                    item["quantity"] -= quantity
                    print(f"Removed {quantity} {item['product'].name}(s) from the cart.")
                    if item["quantity"] == 0:
                        self.items.remove(item)

                else:
                    print("Quantity to remove exceeds the quantity in the cart.")
                return
        print(f"Product with ID {product_id} not found in the cart.")

    def calculate_total(self):
        """Calculate the total price of items in the shopping cart."""
        total = sum(item["product"].price * item["quantity"] for item in self.items)
        return total            
    
class Checkout:
    def __init__(self, shopping_cart):
        self.shopping_cart = shopping_cart
    
    def process_payment(self, payment_amount):
        """Process the payment and complete the checkout."""
        total_amount = self.shopping_cart.calculate_total()

    if payment_amount >= total_amount:
        print(f"Payment successful! Total amount: ${total_amount: .2f}")
        self.shopping_cart.items = [] # Clear the cart after successful payment
    else:
        print("Payment failed. Insufficient funds.")

# Example usage of the shopping cart and checkout system
if __name__ == "__main__":
    # Create some sample products
    laptop = Product(1, "Laptop", 999.99)
    smartphone = Product(2, "Smartphone", 499.99)
    headphones = Product(3, "Headphones", 79.99)

    # Create a shopping cart
    cart = ShoppingCart()

    # Add products to the cart
    cart.add_item(laptop, 2)
    cart.add_item(smartphone, 1)
    cart.add_item(headphones, 3)

    # Display the contents of the cart
    print("Shopping Cart Contents:")
    for item in cart.items:
        print(f"{item['quantity']} x {item['product'].name}")

    # Remove some items from the cart
    cart.remove_item(1, 1)
    cart.remove_item(3, 2)

    # Display the updated contents of the cart
    print("\nUpdated Shopping Cart Contents:")
    for item in cart.items:
        print(f"{item['quantity']} x {item['product'].name}")

    # Create a checkout and process the payment
    checkout = Checkout(cart)
    checkout.process_payment(2000)